plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitLoader.java

```diff
@@ -0,0 +1,262 @@
+import java.util.Set;
+
+import static io.airlift.concurrent.MoreFutures.asVoid;
+import static io.airlift.concurrent.MoreFutures.toListenableFuture;
+import static io.trino.plugin.iceberg.ExpressionConverter.toIcebergExpression;
+import static io.trino.plugin.iceberg.IcebergUtil.deserializePartitionValue;
+import static io.trino.plugin.iceberg.IcebergUtil.getPartitionKeys;
+import static io.trino.plugin.iceberg.IcebergUtil.primitiveFieldTypes;
+import static io.trino.plugin.iceberg.Partition.convertBounds;
+import static io.trino.plugin.iceberg.TypeConverter.toIcebergType;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class IcebergSplitLoader
+        implements ResumableTask
+{
+    private final SchemaTableName schemaTableName;
+    private final Set<IcebergColumnHandle> identityPartitionColumns;
+    private final DynamicFilter dynamicFilter;
+    private final TimeZoneKey timeZoneKey;
+    private final TableScan tableScan;
+    private final AsyncQueue<ConnectorSplit> splitQueue;
+    private final long dynamicFilteringProbeBlockingTimeoutMillis;
+    private final Stopwatch stopwatch;
+
+    private CloseableIterable<CombinedScanTask> combinedScanIterable;
+    private Iterator<FileScanTask> fileScanIterator;
+
+    public IcebergSplitLoader(
+            SchemaTableName schemaTableName,
+            Set<IcebergColumnHandle> identityPartitionColumns,
+            TableScan tableScan,
+            DynamicFilter dynamicFilter,
+            TimeZoneKey timeZoneKey,
+            AsyncQueue<ConnectorSplit> splitQueue,
+            long dynamicFilteringProbeBlockingTimeoutMillis)
+    {
+        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+        this.identityPartitionColumns = requireNonNull(identityPartitionColumns, "identityPartitionColumns is null");
+        this.tableScan = requireNonNull(tableScan, "tableScan is null");
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.timeZoneKey = requireNonNull(timeZoneKey, "timeZoneKey is null");
+        this.splitQueue = requireNonNull(splitQueue, "splitQueue is null");
+        this.dynamicFilteringProbeBlockingTimeoutMillis = dynamicFilteringProbeBlockingTimeoutMillis;
+        this.stopwatch = Stopwatch.createStarted();
+    }
+
+    @Override
+    public TaskStatus process()
+    {
+        long timeLeft = dynamicFilteringProbeBlockingTimeoutMillis - stopwatch.elapsed(MILLISECONDS);
+        if (dynamicFilter.isAwaitable() && timeLeft > 0) {
+            return TaskStatus.continueOn(asVoid(toListenableFuture(dynamicFilter.isBlocked().orTimeout(timeLeft, MILLISECONDS))));
+        }
+
+        if (fileScanIterator == null) {
+            Expression dynamicFilterExpression = toIcebergExpression(dynamicFilter.getCurrentPredicate().transformKeys(IcebergColumnHandle.class::cast));
```

Comment:
Can we shift the `icebergTable.newScan()` part from `IcebergSplitManager#getSplits` here ?
Filter expression should probably be
```
table.getUnenforcedPredicate()
.intersect(dynamicFilter.getCurrentPredicate().transformKeys(IcebergColumnHandle.class::cast))
.simplify(ICEBERG_DOMAIN_COMPACTION_THRESHOLD)
```
This will avoid making iceberg do the AND'ing of filters and also avoid the large IN list failure